### PR TITLE
Enable `errcheck` linter

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

Enable `errcheck` and fix linter errors.

**Link to tracking Issue:** Updates #2789

### CODE REVIEW

File: config/configmap.go [hunk 1]
```diff
@@ -116,7 +116,8 @@ func (l *Map) Set(key string, value interface{}) {
 	// koanf doesn't offer a direct setting mechanism so merging is required.
 	merged := koanf.New(KeyDelimiter)
 	_ = merged.Load(confmap.Provider(map[string]interface{}{key: value}, KeyDelimiter), nil)
-	l.k.Merge(merged)
+	// Ignore error for now; it only fails with strict merge when the types are different
+	_ = l.k.Merge(merged)
```
Comment (mx-psi) [inline config/configmap.go:120]: should Koanf decide to return an error for a different reason we would have no way to report the `Set` failed, so maybe we should return the error here instead?

File: .golangci.yml [hunk 2]
```diff
@@ -138,7 +137,7 @@ issues:
   # The list of ids of default excludes to include or disable. By default it's empty.
   # See the list of default excludes here https://golangci-lint.run/usage/configuration.
   include:
-    - EXC0001
+    # - EXC0001 - errcheck checks that are not usually checked
```
Comment (mx-psi) [inline .golangci.yml:140]: We discussed this on https://github.com/open-telemetry/opentelemetry-collector/pull/2881#discussion_r616834777

File: config/configmap.go (hunk 1 above)
Comment (codeboten) [reply config/configmap.go:120]: seems there's at least a couple of calls that could trigger an error in this func with `Load` just above this line, I would be in favour of returning an error from Set

**Review by codeboten — COMMENTED** [review]
Thanks for doing this, it looks good to me. Will review and approve once a decision on what to do w/ the errors from `Set` is made.

File: config/configgrpc/configgrpc_test.go [hunk 3]
```diff
@@ -40,7 +40,7 @@ import (
 func TestDefaultGrpcClientSettings(t *testing.T) {
 	tt, err := obsreporttest.SetupTelemetry()
 	require.NoError(t, err)
-	defer tt.Shutdown(context.Background())
+	defer func() { require.NoError(t, tt.Shutdown(context.Background())) }()
```
Comment (bogdandrutu) [inline config/configgrpc/configgrpc_test.go:43]: I think t.Cleanup() is nicer.

File: config/configmap.go (hunk 1 above)
Comment (mx-psi) [reply config/configmap.go:120]: Opened #4467 for this, can you assign this to me @codeboten? I will do it on a separate PR to keep this one for non-breaking changes

File: config/configgrpc/configgrpc_test.go (hunk 3 above)
Comment (mx-psi) [reply config/configgrpc/configgrpc_test.go]: Done

**Review by bogdandrutu — APPROVED** [review]

File: obsreport/obsreport_processor.go [hunk 4]
```diff
@@ -78,7 +79,8 @@ func (por *Processor) TracesAccepted(ctx context.Context, numSpans int) {
 // TracesRefused reports that the trace data was refused.
 func (por *Processor) TracesRefused(ctx context.Context, numSpans int) {
```
Comment (bogdandrutu) [inline obsreport/obsreport_processor.go:80]: Same question, should this return error? Does otel return error?

Comment (mx-psi) [reply obsreport/obsreport_processor.go:80]: I don't think otel does as of today (guessing [this](https://pkg.go.dev/go.opentelemetry.io/otel/metric#Int64UpDownCounter.Add) is equivalent for example?). For the obsreport errors I was just following what we do for other components e.g. see https://github.com/open-telemetry/opentelemetry-collector/blob/89e230ef02b469f7f45f9bcdbbfa7ce0f7c0c717/obsreport/obsreport_exporter.go#L108-L114

Comment (bogdandrutu) [reply obsreport/obsreport_processor.go:80]: I saw that, but curios if we should return error because that will be the case when switching to otel.

Comment (bogdandrutu) [reply obsreport/obsreport_processor.go:80]: Seems not to be the case so we are good.

**Review by codeboten — APPROVED** [review]

**Pull request merged**